[iPadOS 26] Web content crash after zooming in, scrolling down and to the right, and opening sidebar in Safari

aprotyas opened the pull request:
#### 33f6aaf56f343a936a38ee4fedd09a82a210b0fe
<pre>
[iPadOS 26] Web content crash after zooming in, scrolling down and to the right, and opening sidebar in Safari
<a href="https://bugs.webkit.org/show_bug.cgi?id=295898">https://bugs.webkit.org/show_bug.cgi?id=295898</a>
<a href="https://rdar.apple.com/153348208">rdar://153348208</a>

Reviewed by Simon Fraser and Wenson Hsieh.

The combination of _overrideLayoutParametersWithMinimumLayoutSize:...
and _setObscuredInsets, say, to present a sidebar UI that lays out web
content in a new (narrower) viewport configuration, can exhibit extreme
page demands under certain conditions.

To understand this bad behavior, let&apos;s mark two time periods:
pre-sidebar presentation (A), and post-sidebar presentation (B).

When a client overrides layout parameters, it triggers layout in the web
content process and leads to resizing of the graphics layers. In some
future layer change commit, the visible rect from the root layer&apos;s
compositing state flush corresponds to the frame&apos;s exposed content rect
in (A). However, the clip rect for child layers, which is zero origin
and layer sized, uses a layer size from (B) -- i.e. narrower by the inset
amount. When zoomed in and far enough to the right, there is no longer
an intersection between these two rects, and so a zero sized visible
rect and layer sized coverage rect is computed. When scrolled down
sufficiently, this coverage rect has a huge height, which means we end
up preparing an enormous number of backing stores, exhibiting jetsams
because of the sheer amount of IOSurfaces we need.

There are a few issues:
1 Both layout parameter override and obscured inset assignment kicks off
  UI process driven update cycles in the web process. The former
  triggers layout and the latter a visible content rect update. Hence,
  whether or not a compositing flush happens with a layer size from (A)
  or (B) is racy. This is not directly fixable because overriding a
  layout parameter leads to multiple layer tree transactions, and we
  cannot reliably block scheduling visible rect updates in the UI
  process while those are done.
2 The frame&apos;s exposed content rect cannot be in (B) until the web
  process performs a layout round. This necessarily resizes layers.
  Which means that there will always be at least one layer change commit
  where we use the outdated exposed content rect from (A) to compute
  coverage rects.

(2) is more damning here, though, because it implies clients using these
APIs in their current shape will necessarily be prone to this bug under
certain conditions.

In this patch, we make an effort to work around the issue highlighted
above. On obscured inset changes, we introduce a fence against layout
size change dispatches until the first layer tree transaction round trip.
We use this round trip to relay the new obscured inset to the web content
process, where the local frame view is informed of changes in the
obscured inset amount across successive visible rect updates. The tile
controller holds on to these deltas to adjust child clip rects as
necessary to avoid the unfounded disjointness of clip rects. More
details inline.

Note that no API test could be added here. The layer change commit with
pre-sidebar exposed content rect still ends up with a correct reconciled
scroll position, so the only behavior to test would be page demand,
which is not meaningfully observed.

* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::clearObscuredInsetsAdjustmentsIfNeeded):
  This provides an explicit path to clear these adjustments to avoid
  an implicit dependency on Page::setObscuredInsets being called
  consecutively with the same values.
(WebCore::LocalFrameView::obscuredInsetsWillChange):
* Source/WebCore/page/LocalFrameView.h:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::setObscuredInsets):
  Compute the diff in the obscured insets from the last visible rect
  update cycle and inform the local frame view accordingly.
* Source/WebCore/platform/RectEdges.h:
(WebCore::operator-):
(WebCore::operator-=):
  Allows for more ergonomic FloatBoxExtent subtractions expression.
* Source/WebCore/platform/graphics/TiledBacking.h:
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
(WebCore::GraphicsLayerCA::computeVisibleAndCoverageRect const):
  If we&apos;re using tiled layers, adjust the clip rect input used for
  coverage computation as appropriate.
* Source/WebCore/platform/graphics/ca/TileController.cpp:
(WebCore::TileController::adjustedTileClipRectForObscuredInsets const):
  The adjustment of the tile clip rect is simply to inflate it by a size
  accounting for the delta in obscured insets across visible rect
  updates. We do this by diffs rather than absolute obscured inset
  values so that the workaround is active and used only when necessary
  (e.g. a new obscured insets are set) .
* Source/WebCore/platform/graphics/ca/TileController.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
  Add a transaction ID `firstTransactionIDAfterObscuredInsetChange` in
  the web view&apos;s per process state. This allows us to hold off on view
  layout size change dispathces till later.
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView _setObscuredInsetsInternal:]):
(-[WKWebView _didCommitLayerTree:]):
  If we&apos;ve committed the layer tree, check if we were waiting for a
  transaction post obscured insets application. If so, stop deferring
  geometry updates.
(-[WKWebView _overrideLayoutParametersWithMinimumLayoutSize:minimumUnobscuredSizeOverride:maximumUnobscuredSizeOverride:]):
(-[WKWebView _shouldDispatchOverridenLayoutParametersImmediately]):
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::updateVisibleContentRects):
* Source/WebKit/WebProcess/WebPage/wc/GraphicsLayerWC.cpp:

Canonical link: <a href="https://commits.webkit.org/297499@main">https://commits.webkit.org/297499@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f1ea6fbd3d9fc4f3bf08e8e4beca95375fb6aabb

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/112022 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/131/builds/31693 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/138/builds/22178 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/118045 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/62259 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/113984 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/130/builds/32369 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/123/builds/40270 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/118045 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/62259 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/114969 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/130/builds/32369 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/138/builds/22178 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/118045 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/130/builds/32369 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/138/builds/22178 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/61897 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/130/builds/32369 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/138/builds/22178 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/121364 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/128/builds/39055 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/123/builds/40270 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/121364 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/121/builds/39436 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/138/builds/22178 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/121364 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/138/builds/22178 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/35077 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/18052 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/127/builds/38949 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/44461 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/38586 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/41914 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/40302 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
```diff
@@ -3079,7 +3088,7 @@ - (UIEdgeInsets)currentlyVisibleContentInsetsWithScale:(CGFloat)scaleFactor obsc
 - (BOOL)_shouldDeferGeometryUpdates
 {
-    return _perProcessState.liveResizeParameters || _perProcessState.dynamicViewportUpdateMode != WebKit::DynamicViewportUpdateMode::NotResizing || _perProcessState.isAnimatingFullScreenExit;
+    return _perProcessState.liveResizeParameters || _perProcessState.dynamicViewportUpdateMode != WebKit::DynamicViewportUpdateMode::NotResizing || _perProcessState.isAnimatingFullScreenExit || _perProcessState.firstTransactionIDAfterInsetAdjustment;
```
Bundling `firstTransactionIDAfterInsetAdjustment` into `_shouldDeferGeometryUpdates` is problematic on at least one count, since it doesn't actually let us get a content rect update across with the _adjusted_ visible rect

whsieh commented on Source/WebKit/Platform/spi/ios/UIKitSPI.h:
```diff
@@ -1280,6 +1280,7 @@ extern const NSString *UIPreviewDataAttachmentIndex;
 extern NSString * const UIPreviewDataAttachmentListIsContentManaged;
 
 UIEdgeInsets UIEdgeInsetsAdd(UIEdgeInsets lhs, UIEdgeInsets rhs, UIRectEdge);
+UIEdgeInsets UIEdgeInsetsSubtract(UIEdgeInsets lhs, UIEdgeInsets rhs, UIRectEdge edges);
```
Is this declaration used anywhere?

aprotyas replied:
Nevermind, this was used in a previous iteration of the patch. No longer needed.

whsieh commented on Source/WebCore/page/Page.cpp:
```diff
@@ -5305,6 +5305,10 @@ void Page::setSceneIdentifier(String&& sceneIdentifier)
 {
+    RefPtr localMainFrame = this->localMainFrame();
+    if (RefPtr view = localMainFrame ? localMainFrame->view() : nullptr)
+        view->obscuredInsetsDidChange(insets - m_obscuredInsets);
```
Would this be a `WillChange` instead of a `DidChange`?

I also wonder if this needs to be called in the case where `m_obscuredInsets == insets`, or if we could put it _after_ the early return.

aprotyas replied:
I'll change it to `WillChange`

aprotyas replied:
Though we need to maintain these calls regardless of the early return because we only want the adjustment for one visible rect update cycle (in the next cycle, this will be a zero sized extent)

smfr commented on Source/WebCore/platform/graphics/ca/TileController.h:
```diff
@@ -288,6 +291,7 @@ class TileController final : public TiledBacking {
     float m_tileDebugBorderWidth { 0 };
     ScrollingModeIndication m_indicatorMode { SynchronousScrollingBecauseOfLackOfScrollingCoordinatorIndication };
     FloatBoxExtent m_obscuredContentInsets;
+    FloatBoxExtent m_activeObscuredInsetsDelta;
```
What's the life time of this? Should it be a std::optional<> and be unset when the state stabilizes?

aprotyas replied:
It's "unset" in the sense that it becomes a zero sized extent when the state stabilizes.

aprotyas replied:
Maybe `std::nullopt` is a better representation.

aprotyas replied:
i.e. instead of `view->obscuredInsetsDidChange(insets - m_obscuredInsets);` we'd call `view->obscuredInsetsDidChange(std::nullopt);` when the inset no longer changes.

aprotyas replied:
(and the lifetime is one visible rect update cycle after unobscured inset assignment)

aprotyas replied:
Made the life time more obvious by using a `std::optional<>` and by providing an explicit unset path `TiledBacking::clearObscuredInsetsAdjustments()`

smfr commented on Source/WebCore/page/LocalFrameView.cpp:
```diff
@@ -1074,6 +1074,18 @@ LayoutPoint LocalFrameView::scrollPositionRespectingCustomFixedPosition() const
 
+void LocalFrameView::clearObscuredInsetsAdjustmentsIfNeeded()
+{
+    if (TiledBacking* tiledBacking = this->tiledBacking())
```
auto

smfr replied:
Might need a smart pointer to keep SaferCPP happy.

aprotyas replied:
True, `CheckedPtr`

smfr commented on Source/WebCore/page/LocalFrameView.cpp:
```diff
@@ -1074,6 +1074,18 @@ LayoutPoint LocalFrameView::scrollPositionRespectingCustomFixedPosition() const
+
+void LocalFrameView::obscuredInsetsWillChange(FloatBoxExtent&& obscuredInsetsDelta)
+{
+    if (TiledBacking* tiledBacking = this->tiledBacking())
```
Ditto

aprotyas committed:
[iPadOS 26] Web content crash after zooming in, scrolling down and to the right, and opening sidebar in Safari
https://bugs.webkit.org/show_bug.cgi?id=295898
rdar://153348208

Reviewed by Simon Fraser and Wenson Hsieh.

The combination of _overrideLayoutParametersWithMinimumLayoutSize:...
and _setObscuredInsets, say, to present a sidebar UI that lays out web
content in a new (narrower) viewport configuration, can exhibit extreme
page demands under certain conditions.

To understand this bad behavior, let's mark two time periods:
pre-sidebar presentation (A), and post-sidebar presentation (B).

When a client overrides layout parameters, it triggers layout in the web
content process and leads to resizing of the graphics layers. In some
future layer change commit, the visible rect from the root layer's
compositing state flush corresponds to the frame's exposed content rect
in (A). However, the clip rect for child layers, which is zero origin
and layer sized, uses a layer size from (B) -- i.e. narrower by the inset
amount. When zoomed in and far enough to the right, there is no longer
an intersection between these two rects, and so a zero sized visible
rect and layer sized coverage rect is computed. When scrolled down
sufficiently, this coverage rect has a huge height, which means we end
up preparing an enormous number of backing stores, exhibiting jetsams
because of the sheer amount of IOSurfaces we need.

There are a few issues:
1 Both layout parameter override and obscured inset assignment kicks off
  UI process driven update cycles in the web process. The former
  triggers layout and the latter a visible content rect update. Hence,
  whether or not a compositing flush happens with a layer size from (A)
  or (B) is racy. This is not directly fixable because overriding a
  layout parameter leads to multiple layer tree transactions, and we
  cannot reliably block scheduling visible rect updates in the UI
  process while those are done.
2 The frame's exposed content rect cannot be in (B) until the web
  process performs a layout round. This necessarily resizes layers.
  Which means that there will always be at least one layer change commit
  where we use the outdated exposed content rect from (A) to compute
  coverage rects.

(2) is more damning here, though, because it implies clients using these
APIs in their current shape will necessarily be prone to this bug under
certain conditions.

In this patch, we make an effort to work around the issue highlighted
above. On obscured inset changes, we introduce a fence against layout
size change dispatches until the first layer tree transaction round trip.
We use this round trip to relay the new obscured inset to the web content
process, where the local frame view is informed of changes in the
obscured inset amount across successive visible rect updates. The tile
controller holds on to these deltas to adjust child clip rects as
necessary to avoid the unfounded disjointness of clip rects. More
details inline.

Note that no API test could be added here. The layer change commit with
pre-sidebar exposed content rect still ends up with a correct reconciled
scroll position, so the only behavior to test would be page demand,
which is not meaningfully observed.

* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::clearObscuredInsetsAdjustmentsIfNeeded):
  This provides an explicit path to clear these adjustments to avoid
  an implicit dependency on Page::setObscuredInsets being called
  consecutively with the same values.
(WebCore::LocalFrameView::obscuredInsetsWillChange):
* Source/WebCore/page/LocalFrameView.h:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::setObscuredInsets):
  Compute the diff in the obscured insets from the last visible rect
  update cycle and inform the local frame view accordingly.
* Source/WebCore/platform/RectEdges.h:
(WebCore::operator-):
(WebCore::operator-=):
  Allows for more ergonomic FloatBoxExtent subtractions expression.
* Source/WebCore/platform/graphics/TiledBacking.h:
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
(WebCore::GraphicsLayerCA::computeVisibleAndCoverageRect const):
  If we're using tiled layers, adjust the clip rect input used for
  coverage computation as appropriate.
* Source/WebCore/platform/graphics/ca/TileController.cpp:
(WebCore::TileController::adjustedTileClipRectForObscuredInsets const):
  The adjustment of the tile clip rect is simply to inflate it by a size
  accounting for the delta in obscured insets across visible rect
  updates. We do this by diffs rather than absolute obscured inset
  values so that the workaround is active and used only when necessary
  (e.g. a new obscured insets are set) .
* Source/WebCore/platform/graphics/ca/TileController.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
  Add a transaction ID `firstTransactionIDAfterObscuredInsetChange` in
  the web view's per process state. This allows us to hold off on view
  layout size change dispathces till later.
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView _setObscuredInsetsInternal:]):
(-[WKWebView _didCommitLayerTree:]):
  If we've committed the layer tree, check if we were waiting for a
  transaction post obscured insets application. If so, stop deferring
  geometry updates.
(-[WKWebView _overrideLayoutParametersWithMinimumLayoutSize:minimumUnobscuredSizeOverride:maximumUnobscuredSizeOverride:]):
(-[WKWebView _shouldDispatchOverridenLayoutParametersImmediately]):
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::updateVisibleContentRects):
* Source/WebKit/WebProcess/WebPage/wc/GraphicsLayerWC.cpp:

Canonical link: https://commits.webkit.org/297499@main
